File: Data Structures & Algorithms/depth-of-binary-tree/submission-6.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int DFS(TreeNode* root)
    {
        if (root == nullptr)
        {
            return 0;
        }

        int leftSubtree = DFS(root->left);
        int rightSubtree = DFS(root->right);

        return 1 + std::max(leftSubtree, rightSubtree); // 1 here to include itself (in addition to the root).
    }

    int maxDepth(TreeNode* root) {
        return DFS(root);
    }
};

File: Data Structures & Algorithms/count-paths/submission-6.py
class Solution:
    def uniquePaths(self, m: int, n: int) -> int:
        leMap = {}
        def DFS(i, j):
            if (i, j) in leMap:
                return leMap[(i, j)]

            if i >= m or j >= n:
                return 0
            
            if i == m - 1 and j == n - 1:
                return 1 

            totalWays = 0
            totalWays += DFS(i + 1, j)
            totalWays += DFS(i, j + 1)

            leMap[(i, j)] = totalWays
            return totalWays
        return DFS(0, 0)
        